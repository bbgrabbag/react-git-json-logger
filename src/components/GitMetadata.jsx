import React from "react";
import { DEFAULT_TEMPLATE_TEST_ID } from "../test-utils/mock.jsx";

/**
 * @param {object} props.metadata git repo info: user, branch, commit, date, remotes
 * @param {(metadata) => JSX.Element} props.render custom render component
 * @returns {JSX.Element} JSX Element
 */
export const GitMetadata = (props) => {
  if (!props.metadata)
    throw Error(
      `Invalid <GitMetadata> props: Must provide 'metadata' attribute`
    );

  if (props.render) return props.render({ metadata: props.metadata });

  return (
    <div
      data-testid={DEFAULT_TEMPLATE_TEST_ID}
      style={{ fontFamily: "monospace", padding: "1rem" }}
    >
      {Object.keys(props.metadata).map((k) =>
        k === "remotes" ? (
          <div key={k}>
            <p>
              <b>{k}</b>:
            </p>
            <ul>
              {props.metadata[k].map((remote, i) => {
                return (
                  <li key={remote + i}>
                    <b>{remote.name}</b>: <span>{remote.url}</span>
                  </li>
                );
              })}
            </ul>
          </div>
        ) : (
          <p key={k}>
            <b>{k}</b>: <span>{props.metadata[k]}</span>
          </p>
        )
      )}
    </div>
  );
};
