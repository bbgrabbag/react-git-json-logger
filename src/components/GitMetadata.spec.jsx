import React from "react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { GitMetadata } from "./GitMetadata";
import {
  CustomTemplate,
  CUSTOM_TEMPLATE_TEST_ID,
  DEFAULT_TEMPLATE_TEST_ID,
  MOCK_METADATA,
} from "../test-utils/mock";

describe("Test <GitMetadata/> Component", () => {
  it(`Should throw with missing props: 'metadata'`, async () => {
    expect(() => render(<GitMetadata />)).toThrowError(
      `Invalid <GitMetadata> props: Must provide 'metadata' attribute`
    );
  });

  it("Should render default template with metadata populated", async () => {
    const cmp = render(<GitMetadata metadata={MOCK_METADATA} />);
    const parent = await cmp.findByTestId(DEFAULT_TEMPLATE_TEST_ID);
    expect(parent).toBeDefined();
    expect(parent).toMatchSnapshot();
  });

  it("Should render custom template with metadata populated", async () => {
    const cmp = render(
      <GitMetadata metadata={MOCK_METADATA} render={CustomTemplate} />
    );
    const parent = await cmp.findByTestId(CUSTOM_TEMPLATE_TEST_ID);
    expect(parent).toBeDefined();
    expect(parent).toMatchSnapshot();
  });
});
