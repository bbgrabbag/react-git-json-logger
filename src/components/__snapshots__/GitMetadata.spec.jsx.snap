// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <GitMetadata/> Component Should render custom template with metadata populated 1`] = `
<div
  data-testid="rgjs__custom-template"
>
  <p>
    <b>
      User
    </b>
    :
    Ben Turner &lt;bbgrabbag@gmail.com&gt;
  </p>
</div>
`;

exports[`Test <GitMetadata/> Component Should render default template with metadata populated 1`] = `
<div
  data-testid="rgjs__default-template"
  style="font-family: monospace; padding: 1rem;"
>
  <p>
    <b>
      user
    </b>
    : 
    <span>
      Ben Turner &lt;bbgrabbag@gmail.com&gt;
    </span>
  </p>
  <p>
    <b>
      branch
    </b>
    : 
    <span>
      master
    </span>
  </p>
  <p>
    <b>
      commit
    </b>
    : 
    <span>
      76a24814698de41e5acf369c65575c83144c4726
    </span>
  </p>
  <p>
    <b>
      date
    </b>
    : 
    <span>
      2022-12-11T02:39:34.000Z
    </span>
  </p>
  <div>
    <p>
      <b>
        remotes
      </b>
      :
    </p>
    <ul>
      <li>
        <b>
          origin
        </b>
        : 
        <span>
          git@github.com:bbgrabbag/react-git-json-logger.git (fetch)
        </span>
      </li>
      <li>
        <b>
          origin
        </b>
        : 
        <span>
          git@github.com:bbgrabbag/react-git-json-logger.git (push)
        </span>
      </li>
    </ul>
  </div>
</div>
`;
